### DISCUSSION TITLE
General question about KB 102501

### DISCUSSION THREAD

**Discussion by ricky7uio** [tabulartools/dax-optimizer#42] [Q&A] [answered]
https://kb.daxoptimizer.com/d/102501

I think this question is for @marcosqlbi. 

In our main model, we have written a lot of measures using the following pattern: 

```
VAR FILTER1 =
    FILTER (
        ALL (
            TABLEA[COLUMN1],
            TABLEA[COLUMN2]
        ),
        TABLEA[COLUMN1] = "XYZ"
            && TABLEA[COLUMN2] = "ABC"
    )
VAR CALCULATION1 =
    CALCULATE (
        [BASE MEASURE 1],
        FILTER1
    )
VAR CALCULATION2 =
    CALCULATE (
        [BASE MEASURE 2],
        FILTER1
    )
RETURN
    //ANY CALCULATION, HERE JUST AN EXAMPLE
    DIVIDE (
        CALCULATION1,
        CALCULATION2
    )


```
This measures are generally flagged with the issue [A FILTER function is used instead of modifying the filter context](https://kb.daxoptimizer.com/d/102501)

We had the idea and I remembered you or Alberto explaining it on one of your video courses that writing this: 

```
CALCULATE (
    [BASE MEASURE 1],
    TABLEA[COLUMN1] = "XYZ",
    TABLEA[COLUMN2] = "ABC"
)
```
Is the same as this: 

```
CALCULATE (
    [BASE MEASURE 1],
    FILTER (
        ALL (
            TABLEA[COLUMN1],
            TABLEA[COLUMN2]
        ),
        TABLEA[COLUMN1] = "XYZ"
            && TABLEA[COLUMN2] = "ABC"
    )
)
```
So that is why we decided to make the measure less redundant and easier to read (in a real one we have more lines of code) and use the patterns shown above. Is our understanding incorrect? 

Just out of curiosity, I've tried replacing what we had in our current pattern to the more verbose code and I was not really able to identify a performance improvement. Also the possible optimizations shown in KB 102501 seem as not really applicable in this case, right? 

Thank you for taking the time for reading and answering my question. We are enjoying and learning very much from the DAXOPTIMIZER tool.

**Comment by marcosqlbi** [comment]
Great question!
Short answer: a table filter with two or more columns may require a materialization that is often not needed for a single-column filter. You described a requirement for an OR condition between two columns, but it's unnecessary and potentially slower (sometimes MUCH slower) for an AND condition.
This:
`CALCULATE (
    [BASE MEASURE 1],
    TABLEA[COLUMN1] = "XYZ",
    TABLEA[COLUMN2] = "ABC"
)`
Corresponds to this:

`CALCULATE (
    [BASE MEASURE 1],
    FILTER ( 
        ALL ( TABLEA[COLUMN1] ), 
        TABLEA[COLUMN1] = "XYZ" 
    ),
    FILTER ( 
        ALL ( TABLEA[COLUMN2] ), 
        TABLEA[COLUMN2] = "ABC"
    )
)`

Thus, in this specific case, the single-column filter is a better option. Indeed, if you try to write this filter in CALCULATE:

`TABLEA[COLUMN1] = "XYZ"
            && TABLEA[COLUMN2] = "ABC"
`
You will get a different issue that suggests splitting the filter in two.

If you have a case where the filter is legit (e.g. for an OR condition), then storing the filter in a table makes sense - if there is an issue you could ignore it in that case, but please send me an example of that so we can evaluate whether to skip the detection in a case that we could consider safe.

Thank you!

**Reply by ricky7uio** [reply]
Thank you for your answer @marcosqlbi. 

I took my time to reply since I tried to wrap my head around your answer. 

So if I understand correctly, it would be better to do something like this: 

```
VAR FILTER1 =
    FILTER (
        ALL ( TABLEA[COLUMN1] ),
        TABLEA[COLUMN1] = "XYZ"
    )
VAR FILTER2 =
    FILTER (
        ALL ( TABLEA[COLUMN2] ),
        TABLEA[COLUMN2] = "ABC"
    )
VAR CALCULATION1 =
    CALCULATE (
        [BASE MEASURE 1],
        FILTER1,
        FILTER2
    )
VAR CALCULATION2 =
    CALCULATE (
        [BASE MEASURE 2],
        FILTER1,
        FILTER2
    )
RETURN
    //ANY CALCULATION, HERE JUST AN EXAMPLE
    DIVIDE (
        CALCULATION1,
        CALCULATION2
    )
```
Which would be exactly the same as doing this: 

```
VAR CALCULATION1 =
    CALCULATE (
        [BASE MEASURE 1],
        TABLEA[COLUMN1] = "XYZ",
        TABLEA[COLUMN2] = "ABC"
    )
VAR CALCULATION2 =
    CALCULATE (
        [BASE MEASURE 2],
        TABLEA[COLUMN1] = "XYZ",
        TABLEA[COLUMN2] = "ABC"
    )
RETURN
    //ANY CALCULATION, HERE JUST AN EXAMPLE
    DIVIDE (
        CALCULATION1,
        CALCULATION2
    )
```
The only part I don't understand is:

> You described a requirement for an OR condition between two columns, 

In  TABLEA[COLUMN1] = "XYZ" && TABLEA[COLUMN2] = "ABC" I think I'm using an AND operator. Maybe I'm missing something. Could you please elaborate a bit on that? 

Thanks again for your time!

Best regards,
Ricardo

**Reply by marcosqlbi** [reply, marked as answer]
Your example is correct - so IMHO there is not much advantage in writing the filters in a variable before CALCULATE.
Regarding OR and AND: if you have two conditions on two different columns with an AND condition (like your &&), then you can split that condition in two different single-column filters, as we did in the example, because all the filters in the filter context are in an AND condition.
However, if you have 
`T[C1] = "ABC" || T[C2] = "XYZ"
`
then you cannot split the condition in two filters, you need a single filter with two columns in this case.
I hope it's clearer now.

**Reply by ricky7uio** [reply]
Thank you Marco. Now it's clear to me!

**Reply by marcosqlbi** [reply]
Great - thank you fo closing the loop!
